{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prediction of TF binding based on DNA physical properties\n",
    "COMP561: Computational Biology Methods and Research"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### About:\n",
    "A method to predict TF binding sites based on DNA physical properties.\n",
    "### Why I chose this project\n",
    "- I chose to do this project (came from within vs being imposed from outside)\n",
    "- There were so many things that I didn't know how to do and I faced many issues (e.g. running time). To solve them, I explored and found out solutions myself, in the forms of Python package, way to efficiently run a tool (Galaxy), and so on. It was an assignment, but I acted very proactively. I learned a lot and had a lot of fun.\n",
    "- I was very immersed when I was doing it: Without noticing, I was working for hours in a row and time flew by. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNA physical properties can affect TF binding\n",
    "Structural properties of DNA have a significant impact on DNA's affinity for proteins\n",
    "- e.g. minor groove width (MGW), Roll, helical twist (HelT), and propeller twist (ProT)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "DNA physical properties can be predicted from the local DNA sequence using tools such as DNAshape: https://academic.oup.com/nar/article/43/D1/D103/2439587"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Identification of binding sites for a given TF is important\n",
    "- e.g. Application: structure-based drug discovery"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ChIP-seq combined with position weight matrix (PWM) can be used to identify TF binding sites\n",
    "- Issue: Sites on DNA that look like they should be bound by TFs but are not in reality (=negative examples)\n",
    "- positive examples = sequences in which TFs are bound"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Objective: To examine whether positive examples could be distinguished from negative examples based on predicted structural properties of their sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Physical properties used\n",
    "- Minor groove width (MGW)\n",
    "- Roll\n",
    "- Helical twist (HelT)\n",
    "- Propeller twist (ProT)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TFs used: TATA box-binding protein (TBP)\n",
    "Rationales\n",
    "- TBP: One of the most well-known and essential TF. Required for correct initiation of transcription and is used by all three eukaryotic RNA polymerases"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Machine learning models used:\n",
    "- K-nearest neighbour (KNN)\n",
    "- Logistic regression"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Project Overview: Steps involved\n",
    "1. Identifying a set of bound and non-bound DNA sequences for a given TF (based on existing experimental data)\n",
    "2. Obtaining the DNA physical properties of each sequence\n",
    "3. Training machine learning classifiers to distinguish between bound and unbound sites."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Identification of a set of bound and non-bound DNA sequences for a given TF \n",
    "Based on existing experimental data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# If pyranges is not installed\n",
    "#import sys\n",
    "#!{sys.executable} -m pip install pyranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pyranges as pr\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "overall_runtime_start = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2366151, 6)\n",
      "   Chom  TFStart   TFEnd TFName  ScoreBS StrandBS\n",
      "0  chr1    10461   10476    UA1     2.02        -\n",
      "1  chr1    10464   10479    UA1     2.26        +\n",
      "2  chr1    16245   16260   CTCF     1.97        -\n",
      "3  chr1    89933   89948    NFY     1.95        -\n",
      "4  chr1    91265   91280   CTCF     1.40        +\n",
      "5  chr1    91419   91434   CTCF     2.07        +\n",
      "6  chr1    91421   91436   CTCF     2.80        -\n",
      "7  chr1    91431   91446   CTCF     1.95        -\n",
      "8  chr1   104986  105001   CTCF     2.83        +\n",
      "9  chr1   138972  138987   CTCF     3.43        -\n"
     ]
    }
   ],
   "source": [
    "# Create a dataframe from the textfile\n",
    "file_name = \"factorbookMotifPos.txt\"  # Contains genomic coordinates of positive examples of the target TFs\n",
    "df = pd.read_csv(file_name, usecols=[1,2,3,4,5,6], \n",
    "                 names=['Chom', 'TFStart', 'TFEnd', 'TFName', 'ScoreBS', 'StrandBS'],\n",
    "                 header=None,\n",
    "                 sep='\\t')\n",
    "\n",
    "# Get an overview of the dataframe\n",
    "print(df.shape)\n",
    "print(df.head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AP1       166122\n",
      "CTCF      158004\n",
      "UAK42     130536\n",
      "SP1        83782\n",
      "ZNF263     75577\n",
      "           ...  \n",
      "UA14         149\n",
      "NR4A1         90\n",
      "UAK32         21\n",
      "UAK44          8\n",
      "UAK53          6\n",
      "Name: TFName, Length: 133, dtype: int64\n",
      "\n",
      "Number of types of transcription factors: 133\n"
     ]
    }
   ],
   "source": [
    "# List out all the transcription factors (TFs)\n",
    "df_TFs = df[\"TFName\"].value_counts()\n",
    "print(df_TFs)\n",
    "df_TFs.to_csv(\"TFs.csv\")  # Create a csv file containing all the TFs\n",
    "print(\"\\nNumber of types of transcription factors:\", df_TFs.shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract relevant rows: the TF x the positive examples\n",
    "The + and - strands in the bedfile is unknown; therefore, will deal only with + strand."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the TF\n",
    "tf = 'TBP'  # 'UA7' 'AP2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(392, 6)\n",
      "       Chom   TFStart     TFEnd TFName  ScoreBS StrandBS\n",
      "4756   chr1   1850728   1850748    TBP     0.79        +\n",
      "14974  chr1   8086379   8086399    TBP     1.14        +\n",
      "14975  chr1   8086381   8086401    TBP     1.32        +\n",
      "23746  chr1  12538132  12538152    TBP     1.25        +\n",
      "24187  chr1  12717648  12717668    TBP     1.39        +\n",
      "30749  chr1  17575548  17575568    TBP     1.34        +\n",
      "40367  chr1  23886002  23886022    TBP     1.13        +\n",
      "48368  chr1  27339350  27339370    TBP     1.20        +\n",
      "48743  chr1  27481630  27481650    TBP     1.43        +\n",
      "52121  chr1  28908396  28908416    TBP     1.74        +\n"
     ]
    }
   ],
   "source": [
    "# Extract rows containing TBP binding sites information\n",
    "df_TF = df[df['TFName'] == tf]\n",
    "df_TF_fwd = df_TF[df_TF['StrandBS'] == '+']\n",
    "print(df_TF_fwd.shape)\n",
    "print(df_TF_fwd.head(10))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read the bed file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(78792, 4)\n",
      "Total # of active regulatory regions: 78792 \n",
      "\n",
      "  Chromosome   Start     End     Name\n",
      "0       chr1  237550  237989   chr1.1\n",
      "1       chr1  521310  521756   chr1.2\n",
      "2       chr1  713702  714675   chr1.3\n",
      "3       chr1  762022  763345   chr1.4\n",
      "4       chr1  805100  805473   chr1.5\n",
      "5       chr1  839802  841023   chr1.6\n",
      "6       chr1  848092  848620   chr1.7\n",
      "7       chr1  856280  856826   chr1.8\n",
      "8       chr1  858297  862058   chr1.9\n",
      "9       chr1  873457  874094  chr1.10\n",
      "      Chromosome      Start        End       Name\n",
      "78782       chrX  154764452  154764876  chrX.2507\n",
      "78783       chrX  154807318  154807679  chrX.2508\n",
      "78784       chrX  154841965  154843013  chrX.2509\n",
      "78785       chrX  154880827  154881309  chrX.2510\n",
      "78786       chrX  154891942  154892376  chrX.2511\n",
      "78787       chrX  155049813  155050189  chrX.2512\n",
      "78788       chrX  155087251  155087687  chrX.2513\n",
      "78789       chrX  155094268  155094730  chrX.2514\n",
      "78790       chrX  155110578  155111395  chrX.2515\n",
      "78791       chrX  155196528  155196958  chrX.2516\n"
     ]
    }
   ],
   "source": [
    "# The bed file: contains active regulatory regions in GM12879 (a cell line derived from lymphoblasts)\n",
    "gr = pr.read_bed(\"wgEncodeRegTfbsClusteredV3.GM12878.merged.bed\")\n",
    "df_bed = gr.df\n",
    "print(df_bed.shape)\n",
    "print(\"Total # of active regulatory regions:\", df_bed.shape[0], \"\\n\")\n",
    "\n",
    "# Observe the top and bottom of the bed file\n",
    "print(df_bed.head(10))\n",
    "print(df_bed.tail(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['chr1', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr2', 'chr20', 'chr21', 'chr22', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chrX']\n",
      "-------------\n",
      "['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chrX']\n"
     ]
    }
   ],
   "source": [
    "# List out all chromosomes\n",
    "list_choms = list(set(df_bed[\"Chromosome\"].to_list()))\n",
    "print(sorted(list_choms))\n",
    "\"\"\"The list was created after inspecting the chromosomes contained in the bed file.\n",
    "The list of chromosomes were created as below for ordering (sorting the list of chromosomes generated directly from the bed\n",
    "file dataframe resulted in an undesirable ordering of chromosomes)\"\"\"\n",
    "print(\"-------------\")\n",
    "list_chrs = ['chr%s' % s for s in range(1,23)]\n",
    "list_chrs.append(\"chrX\")\n",
    "print(list_chrs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting negative examples\n",
    "Extracting genomic coordinates of sequences potentially containing negative examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n",
      "dimension of bed_relev (7110, 4)\n",
      "     Chom     NStart       NEnd UnboundTF\n",
      "0    chr1     237550     237989       TBP\n",
      "1    chr1     521310     521756       TBP\n",
      "2    chr1     713702     714675       TBP\n",
      "3    chr1     762022     763345       TBP\n",
      "4    chr1     805100     805473       TBP\n",
      "..    ...        ...        ...       ...\n",
      "319  chr1  249157159  249158079       TBP\n",
      "320  chr1  249167142  249168812       TBP\n",
      "321  chr1  249199877  249201223       TBP\n",
      "322  chr1  249206342  249206773       TBP\n",
      "323  chr1  249218925  249219295       TBP\n",
      "\n",
      "[324 rows x 4 columns]\n",
      "Running time for chr1: 29.52905011177063\n",
      "====================\n",
      "chr2\n",
      "dimension of bed_relev (5929, 4)\n",
      "     Chom     NStart       NEnd UnboundTF\n",
      "0    chr2      11419      12069       TBP\n",
      "1    chr2      18635      19084       TBP\n",
      "2    chr2     142099     142535       TBP\n",
      "3    chr2     172020     172680       TBP\n",
      "4    chr2     246471     246895       TBP\n",
      "..    ...        ...        ...       ...\n",
      "146  chr2  243027932  243028626       TBP\n",
      "147  chr2  243030186  243031282       TBP\n",
      "148  chr2  243040649  243041009       TBP\n",
      "149  chr2  243153621  243154247       TBP\n",
      "150  chr2  243184789  243185169       TBP\n",
      "\n",
      "[151 rows x 4 columns]\n",
      "Running time for chr2: 24.152405500411987\n",
      "====================\n",
      "chr3\n",
      "dimension of bed_relev (4546, 4)\n",
      "     Chom     NStart       NEnd UnboundTF\n",
      "0    chr3      84528      84964       TBP\n",
      "1    chr3     249642     250011       TBP\n",
      "2    chr3     254070     254527       TBP\n",
      "3    chr3     319893     320825       TBP\n",
      "4    chr3     396036     396415       TBP\n",
      "..    ...        ...        ...       ...\n",
      "342  chr3  197793816  197794140       TBP\n",
      "343  chr3  197807383  197807811       TBP\n",
      "344  chr3  197816578  197816960       TBP\n",
      "345  chr3  197894355  197894870       TBP\n",
      "346  chr3  197899817  197900530       TBP\n",
      "\n",
      "[347 rows x 4 columns]\n",
      "Running time for chr3: 14.933109045028687\n",
      "====================\n",
      "chr4\n",
      "dimension of bed_relev (3352, 4)\n",
      "     Chom     NStart       NEnd UnboundTF\n",
      "0    chr4      29087      29584       TBP\n",
      "1    chr4     124078     125860       TBP\n",
      "2    chr4     154651     156499       TBP\n",
      "3    chr4     236026     236407       TBP\n",
      "4    chr4     238222     238692       TBP\n",
      "..    ...        ...        ...       ...\n",
      "113  chr4  190825601  190826202       TBP\n",
      "114  chr4  190833342  190833967       TBP\n",
      "115  chr4  190861243  190862563       TBP\n",
      "116  chr4  190895190  190895731       TBP\n",
      "117  chr4  190935729  190936269       TBP\n",
      "\n",
      "[118 rows x 4 columns]\n",
      "Running time for chr4: 10.307434558868408\n",
      "====================\n",
      "chr5\n",
      "dimension of bed_relev (3808, 4)\n",
      "     Chom     NStart       NEnd UnboundTF\n",
      "0    chr5      13091      13703       TBP\n",
      "1    chr5      21630      22054       TBP\n",
      "2    chr5      42238      42674       TBP\n",
      "3    chr5      49471      49910       TBP\n",
      "4    chr5      57035      57525       TBP\n",
      "..    ...        ...        ...       ...\n",
      "315  chr5  180671786  180672245       TBP\n",
      "316  chr5  180672491  180674405       TBP\n",
      "317  chr5  180687509  180689387       TBP\n",
      "318  chr5  180700284  180700747       TBP\n",
      "319  chr5  180714191  180714761       TBP\n",
      "\n",
      "[320 rows x 4 columns]\n",
      "Running time for chr5: 12.074646472930908\n",
      "====================\n",
      "chr6\n",
      "dimension of bed_relev (4405, 4)\n",
      "     Chom     NStart       NEnd UnboundTF\n",
      "0    chr6     148248     149088       TBP\n",
      "1    chr6     157133     157591       TBP\n",
      "2    chr6     168805     169270       TBP\n",
      "3    chr6     183657     184081       TBP\n",
      "4    chr6     188714     190168       TBP\n",
      "..    ...        ...        ...       ...\n",
      "200  chr6  170842818  170843476       TBP\n",
      "201  chr6  170861859  170862688       TBP\n",
      "202  chr6  170863099  170864066       TBP\n",
      "203  chr6  170884267  170884704       TBP\n",
      "204  chr6  170892805  170894152       TBP\n",
      "\n",
      "[205 rows x 4 columns]\n",
      "Running time for chr6: 24.84160804748535\n",
      "====================\n",
      "chr7\n",
      "dimension of bed_relev (3675, 4)\n",
      "     Chom     NStart       NEnd UnboundTF\n",
      "0    chr7      41559      41803       TBP\n",
      "1    chr7      64017      64413       TBP\n",
      "2    chr7     148798     150276       TBP\n",
      "3    chr7     302690     304247       TBP\n",
      "4    chr7     362003     362386       TBP\n",
      "..    ...        ...        ...       ...\n",
      "228  chr7  158655905  158657372       TBP\n",
      "229  chr7  158678267  158678718       TBP\n",
      "230  chr7  158766142  158766580       TBP\n",
      "231  chr7  158768442  158768803       TBP\n",
      "232  chr7  158769543  158770131       TBP\n",
      "\n",
      "[233 rows x 4 columns]\n",
      "Running time for chr7: 13.02717113494873\n",
      "====================\n",
      "chr8\n",
      "dimension of bed_relev (2856, 4)\n",
      "     Chom     NStart       NEnd UnboundTF\n",
      "0    chr8     179493     179968       TBP\n",
      "1    chr8     182098     182955       TBP\n",
      "2    chr8     195026     197010       TBP\n",
      "3    chr8     215721     216074       TBP\n",
      "4    chr8     356550     357373       TBP\n",
      "..    ...        ...        ...       ...\n",
      "558  chr8  146245378  146245782       TBP\n",
      "559  chr8  146252060  146252494       TBP\n",
      "560  chr8  146276923  146278338       TBP\n",
      "561  chr8  146291261  146291704       TBP\n",
      "562  chr8  146294852  146295333       TBP\n",
      "\n",
      "[563 rows x 4 columns]\n",
      "Running time for chr8: 9.366945505142212\n",
      "====================\n",
      "chr9\n",
      "dimension of bed_relev (1899, 4)\n",
      "      Chom     NStart       NEnd UnboundTF\n",
      "0     chr9     214472     216017       TBP\n",
      "1     chr9     272204     273320       TBP\n",
      "2     chr9     306939     307372       TBP\n",
      "3     chr9     316573     316973       TBP\n",
      "4     chr9     347669     348093       TBP\n",
      "...    ...        ...        ...       ...\n",
      "1321  chr9  140884717  140885168       TBP\n",
      "1322  chr9  140916698  140917178       TBP\n",
      "1323  chr9  140995436  140995804       TBP\n",
      "1324  chr9  141041921  141042377       TBP\n",
      "1325  chr9  141044283  141044920       TBP\n",
      "\n",
      "[1326 rows x 4 columns]\n",
      "Running time for chr9: 5.1762003898620605\n",
      "====================\n",
      "chr10\n",
      "dimension of bed_relev (3175, 4)\n",
      "      Chom     NStart       NEnd UnboundTF\n",
      "0    chr10     119987     120366       TBP\n",
      "1    chr10     122247     122802       TBP\n",
      "2    chr10     180503     182596       TBP\n",
      "3    chr10     183007     183438       TBP\n",
      "4    chr10     224136     224577       TBP\n",
      "..     ...        ...        ...       ...\n",
      "423  chr10  135310867  135311347       TBP\n",
      "424  chr10  135333447  135334337       TBP\n",
      "425  chr10  135422318  135422750       TBP\n",
      "426  chr10  135460348  135461035       TBP\n",
      "427  chr10  135523048  135523574       TBP\n",
      "\n",
      "[428 rows x 4 columns]\n",
      "Running time for chr10: 10.318393468856812\n",
      "====================\n",
      "chr11\n",
      "dimension of bed_relev (3700, 4)\n",
      "      Chom     NStart       NEnd UnboundTF\n",
      "0    chr11     175748     176263       TBP\n",
      "1    chr11     180103     180673       TBP\n",
      "2    chr11     189571     190445       TBP\n",
      "3    chr11     191712     192557       TBP\n",
      "4    chr11     200233     201092       TBP\n",
      "..     ...        ...        ...       ...\n",
      "452  chr11  134738219  134738855       TBP\n",
      "453  chr11  134812061  134812495       TBP\n",
      "454  chr11  134897382  134897744       TBP\n",
      "455  chr11  134944292  134944716       TBP\n",
      "456  chr11  134945558  134946032       TBP\n",
      "\n",
      "[457 rows x 4 columns]\n",
      "Running time for chr11: 15.068711042404175\n",
      "====================\n",
      "chr12\n",
      "dimension of bed_relev (3828, 4)\n",
      "      Chom     NStart       NEnd UnboundTF\n",
      "0    chr12      77398      77779       TBP\n",
      "1    chr12      83238      83703       TBP\n",
      "2    chr12      94059      94555       TBP\n",
      "3    chr12      95189      95673       TBP\n",
      "4    chr12     175966     176674       TBP\n",
      "..     ...        ...        ...       ...\n",
      "252  chr12  133696593  133697729       TBP\n",
      "253  chr12  133706236  133707544       TBP\n",
      "254  chr12  133732276  133733418       TBP\n",
      "255  chr12  133757761  133758576       TBP\n",
      "256  chr12  133778918  133781392       TBP\n",
      "\n",
      "[257 rows x 4 columns]\n",
      "Running time for chr12: 16.378451824188232\n",
      "====================\n",
      "chr13\n",
      "dimension of bed_relev (1547, 4)\n",
      "      Chom     NStart       NEnd UnboundTF\n",
      "0    chr13   19663431   19663859       TBP\n",
      "1    chr13   19728205   19728648       TBP\n",
      "2    chr13   20013211   20013821       TBP\n",
      "3    chr13   20138903   20139341       TBP\n",
      "4    chr13   20160850   20162182       TBP\n",
      "..     ...        ...        ...       ...\n",
      "238  chr13  115000007  115001175       TBP\n",
      "239  chr13  115045868  115047881       TBP\n",
      "240  chr13  115078504  115079308       TBP\n",
      "241  chr13  115079323  115080479       TBP\n",
      "242  chr13  115105257  115105883       TBP\n",
      "\n",
      "[243 rows x 4 columns]\n",
      "Running time for chr13: 4.419226408004761\n",
      "====================\n",
      "chr14\n",
      "dimension of bed_relev (2102, 4)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      Chom     NStart       NEnd UnboundTF\n",
      "0    chr14   20264672   20265133       TBP\n",
      "1    chr14   20346179   20346611       TBP\n",
      "2    chr14   20415835   20416217       TBP\n",
      "3    chr14   20612067   20612495       TBP\n",
      "4    chr14   20664051   20664481       TBP\n",
      "..     ...        ...        ...       ...\n",
      "519  chr14  107256518  107256971       TBP\n",
      "520  chr14  107259096  107260261       TBP\n",
      "521  chr14  107280756  107281564       TBP\n",
      "522  chr14  107283692  107284122       TBP\n",
      "523  chr14  107287341  107287792       TBP\n",
      "\n",
      "[524 rows x 4 columns]\n",
      "Running time for chr14: 5.929144620895386\n",
      "====================\n",
      "chr15\n",
      "dimension of bed_relev (1375, 4)\n",
      "       Chom     NStart       NEnd UnboundTF\n",
      "0     chr15   20398188   20398853       TBP\n",
      "1     chr15   20561391   20562017       TBP\n",
      "2     chr15   21183357   21184021       TBP\n",
      "3     chr15   21927229   21927614       TBP\n",
      "4     chr15   22387307   22388087       TBP\n",
      "...     ...        ...        ...       ...\n",
      "1177  chr15  102215983  102216558       TBP\n",
      "1178  chr15  102264074  102264866       TBP\n",
      "1179  chr15  102313914  102314376       TBP\n",
      "1180  chr15  102351126  102351488       TBP\n",
      "1181  chr15  102417810  102418300       TBP\n",
      "\n",
      "[1182 rows x 4 columns]\n",
      "Running time for chr15: 3.1824891567230225\n",
      "====================\n",
      "chr16\n",
      "dimension of bed_relev (2690, 4)\n",
      "      Chom    NStart      NEnd UnboundTF\n",
      "0    chr16     60685     61248       TBP\n",
      "1    chr16     88736     89619       TBP\n",
      "2    chr16    102799    104328       TBP\n",
      "3    chr16    115628    116022       TBP\n",
      "4    chr16    126161    126597       TBP\n",
      "..     ...       ...       ...       ...\n",
      "286  chr16  90085855  90086456       TBP\n",
      "287  chr16  90088582  90089266       TBP\n",
      "288  chr16  90144111  90144684       TBP\n",
      "289  chr16  90148092  90149307       TBP\n",
      "290  chr16  90172525  90173041       TBP\n",
      "\n",
      "[291 rows x 4 columns]\n",
      "Running time for chr16: 9.193415880203247\n",
      "====================\n",
      "chr17\n",
      "dimension of bed_relev (3461, 4)\n",
      "      Chom    NStart      NEnd UnboundTF\n",
      "0    chr17     13260     13812       TBP\n",
      "1    chr17     14276     14706       TBP\n",
      "2    chr17     65616     66238       TBP\n",
      "3    chr17    159079    159716       TBP\n",
      "4    chr17    235363    235699       TBP\n",
      "..     ...       ...       ...       ...\n",
      "308  chr17  81034792  81035393       TBP\n",
      "309  chr17  81037142  81037801       TBP\n",
      "310  chr17  81039087  81039348       TBP\n",
      "311  chr17  81122414  81122976       TBP\n",
      "312  chr17  81152946  81153922       TBP\n",
      "\n",
      "[313 rows x 4 columns]\n",
      "Running time for chr17: 13.895840883255005\n",
      "====================\n",
      "chr18\n",
      "dimension of bed_relev (1142, 4)\n",
      "      Chom    NStart      NEnd UnboundTF\n",
      "0    chr18     10035     10832       TBP\n",
      "1    chr18     11570     12026       TBP\n",
      "2    chr18     63492     63976       TBP\n",
      "3    chr18     73279     73887       TBP\n",
      "4    chr18    158011    159327       TBP\n",
      "..     ...       ...       ...       ...\n",
      "456  chr18  77958109  77958490       TBP\n",
      "457  chr18  77958581  77958952       TBP\n",
      "458  chr18  77960106  77960481       TBP\n",
      "459  chr18  78005435  78005812       TBP\n",
      "460  chr18  78015690  78016294       TBP\n",
      "\n",
      "[461 rows x 4 columns]\n",
      "Running time for chr18: 2.5561649799346924\n",
      "====================\n",
      "chr19\n",
      "dimension of bed_relev (3360, 4)\n",
      "      Chom    NStart      NEnd UnboundTF\n",
      "0    chr19    267332    267784       TBP\n",
      "1    chr19    277562    277920       TBP\n",
      "2    chr19    281938    282379       TBP\n",
      "3    chr19    291183    291567       TBP\n",
      "4    chr19    295163    295783       TBP\n",
      "..     ...       ...       ...       ...\n",
      "212  chr19  59083849  59085248       TBP\n",
      "213  chr19  59086474  59087538       TBP\n",
      "214  chr19  59090007  59090287       TBP\n",
      "215  chr19  59092477  59092961       TBP\n",
      "216  chr19  59117568  59118300       TBP\n",
      "\n",
      "[217 rows x 4 columns]\n",
      "Running time for chr19: 12.710009098052979\n",
      "====================\n",
      "chr20\n",
      "dimension of bed_relev (1705, 4)\n",
      "      Chom    NStart      NEnd UnboundTF\n",
      "0    chr20    118345    118776       TBP\n",
      "1    chr20    189572    189997       TBP\n",
      "2    chr20    200139    200589       TBP\n",
      "3    chr20    216489    216919       TBP\n",
      "4    chr20    239552    240009       TBP\n",
      "..     ...       ...       ...       ...\n",
      "588  chr20  62823725  62824162       TBP\n",
      "589  chr20  62854747  62855197       TBP\n",
      "590  chr20  62870108  62870873       TBP\n",
      "591  chr20  62886583  62887545       TBP\n",
      "592  chr20  62958750  62959393       TBP\n",
      "\n",
      "[593 rows x 4 columns]\n",
      "Running time for chr20: 5.395610332489014\n",
      "====================\n",
      "chr21\n",
      "dimension of bed_relev (468, 4)\n",
      "      Chom    NStart      NEnd UnboundTF\n",
      "0    chr21   9484957   9485598       TBP\n",
      "1    chr21   9584754   9585228       TBP\n",
      "2    chr21   9590106   9590558       TBP\n",
      "3    chr21   9909494   9910106       TBP\n",
      "4    chr21   9968064   9968900       TBP\n",
      "..     ...       ...       ...       ...\n",
      "384  chr21  47877830  47879263       TBP\n",
      "385  chr21  47924158  47924518       TBP\n",
      "386  chr21  48054869  48056185       TBP\n",
      "387  chr21  48059635  48060093       TBP\n",
      "388  chr21  48081060  48081441       TBP\n",
      "\n",
      "[389 rows x 4 columns]\n",
      "Running time for chr21: 1.0691413879394531\n",
      "====================\n",
      "chr22\n",
      "dimension of bed_relev (1529, 4)\n",
      "      Chom    NStart      NEnd UnboundTF\n",
      "0    chr22  17255423  17255850       TBP\n",
      "1    chr22  17264720  17265356       TBP\n",
      "2    chr22  17372730  17373278       TBP\n",
      "3    chr22  17392110  17392736       TBP\n",
      "4    chr22  17398288  17398677       TBP\n",
      "..     ...       ...       ...       ...\n",
      "163  chr22  51156410  51156834       TBP\n",
      "164  chr22  51170409  51171139       TBP\n",
      "165  chr22  51195433  51195814       TBP\n",
      "166  chr22  51213530  51214088       TBP\n",
      "167  chr22  51221775  51222404       TBP\n",
      "\n",
      "[168 rows x 4 columns]\n",
      "Running time for chr22: 4.418184757232666\n",
      "====================\n",
      "chrX\n",
      "dimension of bed_relev (2329, 4)\n",
      "     Chom     NStart       NEnd UnboundTF\n",
      "0    chrX     199693     200486       TBP\n",
      "1    chrX     283862     285223       TBP\n",
      "2    chrX     285436     285866       TBP\n",
      "3    chrX     334584     334968       TBP\n",
      "4    chrX     346862     348533       TBP\n",
      "..    ...        ...        ...       ...\n",
      "194  chrX  155049813  155050189       TBP\n",
      "195  chrX  155087251  155087687       TBP\n",
      "196  chrX  155094268  155094730       TBP\n",
      "197  chrX  155110578  155111395       TBP\n",
      "198  chrX  155196528  155196958       TBP\n",
      "\n",
      "[199 rows x 4 columns]\n",
      "Running time for chrX: 5.988030433654785\n",
      "====================\n",
      "      Chom     NStart       NEnd UnboundTF\n",
      "0     chr1     237550     237989       TBP\n",
      "1     chr1     521310     521756       TBP\n",
      "2     chr1     713702     714675       TBP\n",
      "3     chr1     762022     763345       TBP\n",
      "4     chr1     805100     805473       TBP\n",
      "...    ...        ...        ...       ...\n",
      "9304  chrX  155049813  155050189       TBP\n",
      "9305  chrX  155087251  155087687       TBP\n",
      "9306  chrX  155094268  155094730       TBP\n",
      "9307  chrX  155110578  155111395       TBP\n",
      "9308  chrX  155196528  155196958       TBP\n",
      "\n",
      "[9309 rows x 4 columns]\n",
      "(9309, 4)\n",
      "Total running time:  253.97415351867676\n"
     ]
    }
   ],
   "source": [
    "## Before the edit on df_TF_negative\n",
    "start_time_total = time.time()\n",
    "df_TF_negative_all = pd.DataFrame(columns=['Chom','NStart','NEnd','UnboundTF'])  # An empty dataframe used later\n",
    "\n",
    "for chom in list_chrs:  # For each chromosome (chr)\n",
    "    start_time = time.time()\n",
    "    print(chom)\n",
    "\n",
    "    # Extract portion of the bed file corresponding to the chromosome\n",
    "    df_bed_chom = df_bed[df_bed[\"Chromosome\"] == chom]\n",
    "    # print(df_bed_chom.head(10))\n",
    "    # Extract portion of the genomic coordinate (of +'ve example) df corresponding to the chr\n",
    "    df_TF_chom = df_TF_fwd[df_TF_fwd[\"Chom\"] == chom]\n",
    "    #print(df_TF_chom.head(10))\n",
    "    \n",
    "    # Sorting to ensure the dataframe is sorted\n",
    "    df_bed_chom_s = df_bed_chom.sort_values(\"Start\")\n",
    "    df_TF_chom_s = df_TF_chom.sort_values(\"TFStart\")\n",
    "\n",
    "    \n",
    "    # For reducing running time\n",
    "    # Get range for the coordinates of active regulatory regions\n",
    "    min_bed = df_bed_chom_s[\"Start\"].min()\n",
    "    max_bed = df_bed_chom_s[\"End\"].max() \n",
    "\n",
    "   # Get range for the coordinates of TF bound sites (positive examples)\n",
    "    min_TF = df_TF_chom_s[\"TFStart\"].min()\n",
    "    max_TF = df_TF_chom_s[\"TFEnd\"].max()\n",
    "\n",
    "    \"\"\"\n",
    "    Idea:\n",
    "    Negative examples will fall under the active regulatory region.\n",
    "    => Potential negative examples will be regions in active regulatory regions that \n",
    "    doesn't loverlap with locations of positive examples\n",
    "    i.e. Negative examples can't possibly be anywhere in the active regulatory region\n",
    "    that overlaps with positive examples\n",
    "    \n",
    "    min active      minTF                 maxTF            max active\n",
    "    \n",
    "    |=================|----------------------|===============|\n",
    "    \"\"\"\n",
    "    # Extract all the coordinates of active regulatory region that comes before the \n",
    "    # smallest coordinate among the positive examples for the chromosome\n",
    "    df_bed_non_top = df_bed_chom_s[df_bed_chom_s[\"End\"] < min_TF]\n",
    "    df_bed_non_top = df_bed_non_top.rename(columns = {\"Chromosome\":\"Chom\", \"Start\":\"NStart\", \"End\":\"NEnd\"})\n",
    "    if df_bed_non_top.empty == False:\n",
    "        df_bed_non_top.loc[:, \"UnboundTF\"] = tf\n",
    "    df_bed_non_top = df_bed_non_top.drop([\"Name\"], axis=1)\n",
    "\n",
    "    # Extract all the coordinates of active regulatory region that comes after the \n",
    "    # largest coordinate among the positive examples for the chromosome\n",
    "    df_bed_non_bottom = df_bed_chom_s[max_TF < df_bed_chom_s[\"Start\"]] # need to append these\n",
    "    df_bed_non_bottom = df_bed_non_bottom.rename(columns = {\"Chromosome\":\"Chom\", \"Start\":\"NStart\", \"End\":\"NEnd\"})\n",
    "    if df_bed_non_bottom.empty == False:\n",
    "        df_bed_non_bottom.loc[:, \"UnboundTF\"] = tf\n",
    "    df_bed_non_bottom = df_bed_non_bottom.drop([\"Name\"], axis=1)\n",
    "    \n",
    "    \n",
    "    \"\"\"\n",
    "    min active      minTF                 maxTF            max active\n",
    "    \n",
    "    |=================|----------------------|===============|\n",
    "    \"\"\"\n",
    "    # Extract coordinates of active regulatory region that may overlap with coordinates of positive examples\n",
    "    df_bed_relev = df_bed_chom_s[df_bed_chom_s[\"Start\"] <= max_TF]\n",
    "    df_bed_relev = df_bed_relev[min_TF <= df_bed_relev[\"End\"]]\n",
    "    print(\"dimension of bed_relev\", df_bed_relev.shape)\n",
    "    \n",
    "    \n",
    "    # Identify potential regative examples\n",
    "    list_cols_TF = list(df_TF_chom_s.columns.values)\n",
    "    df_TF_negative = pd.DataFrame(columns=['Chom','NStart','NEnd','UnboundTF'])  # Create an empty data frame\n",
    "\n",
    "    count = 0\n",
    "    for index_b, row_b in df_bed_relev.iterrows(): # For each active regulatory region (shortlisted)\n",
    "        X = row_b[\"Start\"]\n",
    "        Y = row_b[\"End\"]\n",
    "\n",
    "        df_TF_Inregion = pd.DataFrame(columns=list_cols_TF)  # Create an empty dataframe\n",
    "        # Gather any positive examples that may fall within the active regulatory region\n",
    "        \"\"\"\n",
    "        X                           Y\n",
    "        |======aaaa===aaaa====aaaa==|\n",
    "        \"\"\"\n",
    "        for index_TF, row_TF in df_TF_chom_s.iterrows():\n",
    "            start = row_TF['TFStart'] \n",
    "            end = row_TF['TFEnd']    \n",
    "            if X <= start and end <= Y:  # If the coordinates of positive examples fall within the active regulatory region\n",
    "                df_1row = df_TF_chom_s.loc[[index_TF]]\n",
    "                df_TF_Inregion = df_TF_Inregion.append(df_1row)\n",
    "\n",
    "        # If at least one positive example falls under the active regulatory region\n",
    "        if df_TF_Inregion.empty == False:\n",
    "            # print(count)\n",
    "            df_TF_Inregion_r = df_TF_Inregion.reset_index()\n",
    "\n",
    "            idx_count = df_TF_Inregion_r.shape[0]\n",
    "\n",
    "            for index_InR, row_InR in df_TF_Inregion_r.iterrows():\n",
    "                st = row_InR[\"TFStart\"]\n",
    "                ed = row_InR[\"TFEnd\"]\n",
    "\n",
    "                if index_InR == 0:  # If it's the first row\n",
    "                    new_row1 = {'Chom': chom, 'NStart': X, 'NEnd': st-1, 'UnboundTF': tf}\n",
    "                    df_TF_negative = df_TF_negative.append(new_row1, ignore_index=True)\n",
    "\n",
    "                if index_InR == idx_count - 1: # If it's the last row\n",
    "                    new_row2 = {'Chom': chom, 'NStart': ed+1, 'NEnd': Y, 'UnboundTF': tf}\n",
    "                    df_TF_negative = df_TF_negative.append(new_row2, ignore_index=True)\n",
    "                    break\n",
    "                else:  # If there are multiple binding sites within a region and if the row is not the last row\n",
    "                    st_2 = int(df_TF_Inregion_r.loc[[index_InR+1]].TFStart)\n",
    "                    ed_2 = int(df_TF_Inregion_r.loc[[index_InR+1]].TFEnd)\n",
    "                    if st_2 > ed: # If the TF binding sites don't overlap\n",
    "                        new_row = {'Chom': chom, 'NStart': ed+1, 'NEnd': st_2-1, 'UnboundTF': tf}\n",
    "                        df_TF_negative = df_TF_negative.append(new_row, ignore_index=True)\n",
    "\n",
    "        count += 1    \n",
    "    \n",
    "    # Create a dataframe that contains all the negative examples\n",
    "    \"\"\"\n",
    "    min active      minTF                      maxTF            max active\n",
    "    |=================|======aaaa===aaaa====aaaa==|===============|\n",
    "    \"\"\"\n",
    "    all_negative_dfs = [df_bed_non_top, df_TF_negative, df_bed_non_bottom]\n",
    "    df_negative_all = pd.concat(all_negative_dfs).reset_index(drop=True)\n",
    "    print(df_negative_all)\n",
    "    end_time = time.time()\n",
    "    run_time = end_time - start_time\n",
    "    print(\"Running time for {}: {}\".format(chom, run_time))\n",
    "    \n",
    "    frames = [df_TF_negative_all, df_negative_all]\n",
    "    df_TF_negative_all = pd.concat(frames).reset_index(drop=True)\n",
    "    \n",
    "    print(\"====================\")\n",
    "\n",
    "end_time_total = time.time()\n",
    "run_time_total = end_time_total - start_time_total\n",
    "print(df_TF_negative_all)\n",
    "print(df_TF_negative_all.shape)\n",
    "print(\"Total running time: \", run_time_total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Chom  NStart    NEnd UnboundTF Strand\n",
      "0  chr1  237550  237989       TBP      +\n",
      "1  chr1  521310  521756       TBP      +\n",
      "2  chr1  713702  714675       TBP      +\n",
      "3  chr1  762022  763345       TBP      +\n",
      "4  chr1  805100  805473       TBP      +\n",
      "5  chr1  839802  841023       TBP      +\n",
      "6  chr1  848092  848620       TBP      +\n",
      "7  chr1  856280  856826       TBP      +\n",
      "8  chr1  858297  862058       TBP      +\n",
      "9  chr1  873457  874094       TBP      +\n",
      "(9309, 5)\n"
     ]
    }
   ],
   "source": [
    "df_TF_negative_fwd = df_TF_negative_all\n",
    "df_TF_negative_all_fwd = df_TF_negative_all.reindex(columns= ['Chom', 'NStart', 'NEnd','UnboundTF', 'Strand'])\n",
    "df_TF_negative_all_fwd.loc[:, \"Strand\"] = \"+\"  # Extract the forward strand\n",
    "print(df_TF_negative_all_fwd.head(10))\n",
    "print(df_TF_negative_all_fwd.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = df_TF_negative_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the result (genomic coordinates of potential negative examples on forward strand) in a csv file\n",
    "fname = tf + \"_negative_fwd.csv\"\n",
    "df_TF_negative_all_fwd.to_csv(fname, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the result in a bed file\n",
    "print(df_TF_negative_all_fwd.head(10))\n",
    "df_TF_negative_all_fwd = df_TF_negative_all_fwd.drop([\"UnboundTF\", \"Strand\"], axis=1)\n",
    "print(df_TF_negative_all_fwd.head(10))\n",
    "fOutName = tf + \"_fwd_neg.bed\"\n",
    "df_TF_negative_all_fwd.to_csv(fOutName, index=False, header=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check that the bed file was successfully created\n",
    "gr2 = pr.read_bed(fOutName)\n",
    "df2_bed = gr2.df\n",
    "print(df2_bed.shape)\n",
    "print(df2_bed.head(5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Obtaining DNA physical properties of each sequence\n",
    "GB Shape: http://rohsdb.cmb.usc.edu/\n",
    "\"Predicted structural properties for every human genome positions are available here\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reminder: Goal of the project\n",
    "To determine whether the sites that are bound (positive examples) can be distinguished from those that are not bound (negative example) based on predicted physical properties of the sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create textfiles that can be directly used as input for DNAshape\n",
    "# Positive examples\n",
    "df_TF_fwd_m = df_TF_fwd.drop([\"TFName\", \"ScoreBS\", \"StrandBS\"], axis=1)\n",
    "print(df_TF_fwd_m.head(10))\n",
    "fout_name = \"factorbookMotifPos_GBshape_\" + tf\n",
    "df_TF_fwd_m.to_csv(fout_name + \".txt\", sep=\"\\t\", index=False, header=False)\n",
    "df_TF_fwd_m.to_csv(fout_name + \".bed\", sep=\"\\t\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "overall_runtime_end = time.time()\n",
    "overall_runtime = overall_runtime_end - overall_runtime_start\n",
    "print(\"Overall running time (s):\", overall_runtime)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1: Extract, from hg19, the actual sequences of bound and unbound sites."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extract sequences of positive examples and potential negative examples: Using Galaxy\n",
    "Workflow\n",
    "1. For each fasta file of a chromosome (e.g. chr1)\n",
    "2. Run Galaxy's \"Extract Genomic DNA\" using the fasta file and \"factorbookMotifPos_TBP_fwd_pos.bed\" which contains the positions of positive examples (where TFs are actually bound)\n",
    "3. Download the output\n",
    "4. Rename the output as, \"seqs_chr#_TBP_fwd_pos.fasta\" (e.g. seqs_chr1_TBP_fwd_pos.fasta)\n",
    "5. Repeat for all chromosome\n",
    "6. Repeat 1-5 for negative example. Change file names as necessary."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
